README.rst

```diff
@@ -226,6 +225,40 @@ If you use multiple Python versions, you can run ``pip-sync`` as
 to upgrade those packages.
 
 
+Should I commit ``requirements.in`` and ``requirements.txt`` to source control?
+===============================================================================
+
+Generally, yes. If you want a reproducible environment installation available from your source control,
+they yes, you should commit both ``requirements.in`` and ``requirements.txt`` to source control.
+
+Note that if you are deploying on multiple Python environments (read the section below),
+then you must commit a seperate output file for each Python environment.
+We suggest to use the ``{env}-requirements.txt`` format
+(ex: ``win32-py2.7-requirements.txt``, ``macos-py3.6-requirements.txt``, etc.).
+
+
+Cross-environment usage of ``requirements.in``/``requirements.txt`` and ``pip-compile``
+=======================================================================================
+
+The dependencies of a package can change depending on the Python environment in which it is installed.
+Here, we define a Python environment as the combination of Operating System, Python version (2.7, 3.6, etc.),
+and Python implementation (CPython, PyPy, etc.). For an exact definition, refer to the possible combinations of
+`PEP 508 environment markers`_.
+
+As the resulting ``requirements.txt`` can differ for each environment, users must execute ``pip-compile``
```

Comment:
I would soften this to "may want to". I would imagine the vast majority of developers these days are using flows where they develop on OS X but deploy on Linux.

While it's technically true that there may be issues, it's very, very uncommon for these issues to actually arise. We have a pretty broad array of web, data engineering, data science, and machine learning projects, and this has only come up for a single project for us – and it's a project that has a sufficiently complicated deploy routine that we're definitely in a "warranties are void" place.

I would instead downgrade this to something like "what if my dependencies behave differently on different platforms?".

While in principle this can be a problem whenever people work across multiple platforms, in practice it's very seldom an actual problem.

Reply:
> While it's technically true that there may be issues, it's very, very uncommon for these issues to actually arise.

My experience contributing and offering support with `pip-tools` and `pipenv`, which reached a wider part of Python users, was quite the opposite. OS mismatch issues did happen often enough to be a pain, and then there were the **Python version based mismatch** which are surprisingly common. If you feel like it, take a dive in `pip-tools` and `pipenv` past issues: its and adventure 😁 .

I expect that the number of libs that will only support py3 will increase over time as we're getting closer to 2020, which I also expect to increase the number of cases of difference in requirements between Python versions as the Py2-Py3 transition continues.

I'm not confortable switching this as a "what-if" warning. I think that the statement a few lines below stated the possibility of reusing the same `requirement.txt` across environments, with the pitfalls. Though, I can add information about what's expected to be more or less risky  (Py2/Py3 vs OS X/Linux vs package tightly bound to the OS, etc.) to avoid being too alarming.

Reply:
I've seen (and contributed) to a number of those issues.

On OS issues, my impression and experience is that they come from a small, vocal minority. If there's an overwhelming majority of users for whom everything "just works", you're unlikely to hear from them. My impression is that the majority of users of pip-tools (and Pipenv) are more in the web dev branch of Python development.

In those cases, where you're mostly just spinning up a Django or Flask app with attendant other dependencies, the OS compatibility issues really do almost never come up. And it really is ubiquitous in that environment to develop on OS X and deploy on Linux.

Python version mismatches are terrible, though, and... yeah, those cases should call for different "compiled" requirements files.